lib/ex_doc/formatter/epub.ex

```diff
@@ -0,0 +1,231 @@
+    File.write("#{output}/OEBPS/content.opf", content)
+  end
+
+  defp generate_toc(output, config, modules, exceptions, protocols, uuid) do
+    content = Templates.toc_template(config, modules ++ exceptions ++ protocols, uuid)
+    File.write("#{output}/OEBPS/toc.ncx", content)
+  end
+
+  defp generate_nav(output, config, modules, exceptions, protocols) do
+    content = Templates.nav_template(config, modules ++ exceptions ++ protocols)
+    File.write("#{output}/OEBPS/nav.html", content)
+  end
+
+  defp generate_title(output, config) do
+    content = Templates.title_template(config)
+    File.write("#{output}/OEBPS/title.html", content)
+  end
+
+  defp generate_list(output, config, nodes) do
+    nodes
+    |> Enum.map(&Task.async(fn -> generate_module_page(output, config, &1) end))
+    |> Enum.map(&Task.await/1)
+  end
+
+  defp generate_epub(output, config) do
+    output = Path.expand(output)
+    target_path =
+      "#{output}/#{config.project}-v#{config.version}.epub"
+      |> String.to_char_list
+
+    {:ok, zip_path} = :zip.create(target_path,
+                                  files_to_add(output),
+                                  compress: ['.css', '.html', '.ncx', '.opf',
+                                             '.jpg', '.png', '.xml'])
+    {:ok, zip_path}
+  end
+
+  defp delete_extras(output) do
+    for target <- ["META-INF", "mimetype", "OEBPS"] do
+      File.rm_rf! "#{output}/#{target}"
+    end
+    :ok
+  end
+
+  ## Helpers
+
+  defp assets do
+   [
+     {"css/*.css", "OEBPS/css" },
+     {"assets/*.xml", "META-INF" },
+     {"assets/mimetype", "." }
+   ]
+  end
+
+  defp files_to_add(path) do
+    File.cd! path, fn ->
+      meta = Path.wildcard("META-INF/*")
+      oebps = Path.wildcard("OEBPS/**/*")
+
+      Enum.reduce meta ++ oebps ++ ["mimetype"], [], fn(f, acc) ->
+        case File.read(f) do
+          {:ok, bin} ->
+            [{f |> String.to_char_list, bin}|acc]
+          {:error, _} ->
+            acc
+        end
+      end
+    end
+  end
+
+  # Helper to format Erlang datetime tuple
+  defp format_datetime do
+    {{year, month, day}, {hour, min, sec}} = :calendar.universal_time()
+    list = [year, month, day, hour, min, sec]
+    :io_lib.format("~4..0B-~2..0B-~2..0BT~2..0B:~2..0B:~2..0BZ", list)
+    |> IO.iodata_to_binary
+  end
+
+  defp generate_module_page(output, config, node) do
+    content =
+      Templates.module_page(config, node)
+      |> valid_xhtml_ids
+    File.write("#{output}/OEBPS/modules/#{node.id}.html", content)
+  end
+
+  defp href(%URI{fragment: nil} = link), do: to_string link
+  defp href(%URI{fragment: fragment} = link) do
+    fragment =
+      fragment
+      |> id_replace
+
+    link
+    |> struct([fragment: fragment])
+    |> to_string
+  end
+
+  defp id_replace(id) do
+    pattern = ~r/[^A-Za-z0-9_.-]/
+    replacement = "--"
+
+    String.replace(id, pattern, replacement)
+  end
```

Comment:
@eksperimental Yes, I think that an ID collision is possible with this workaround, as far as I remember I tried with `EEx` and there was no ID collisions at all, but I can review this tonight with the `Elixir`.

Thinking more about this ID transformation process, IMHO the best place to put this is in the `Autolink` module, we can have an `strict` mode (required by XHTML) to control the kind of IDs that we produce.

The only restrictions of the `id` element in [HTML5](http://www.w3.org/TR/html5/dom.html#the-id-attribute) are the following: Must be unique (duh!), must have at least one character and does not allow space characters. But for XHTML there are more restrictions and *epubchecker* see this kind of violations as errors.

A valid [XML name](http://www.w3.org/TR/REC-xml/#NT-Name) must comply with the following rules:

```
Name             ::=    NameStartChar (NameChar)*
NameStartChar    ::=    ":" | [A-Z] | "_" | [a-z] | [#xC0-#xD6] | [#xD8-#xF6] | [#xF8-#x2FF] | [#x370-#x37D] | [#x37F-#x1FFF] | [#x200C-#x200D] | [#x2070-#x218F] | [#x2C00-#x2FEF] | [#x3001-#xD7FF] | [#xF900-#xFDCF] | [#xFDF0-#xFFFD] | [#x10000-#xEFFFF]
NameChar         ::=    NameStartChar | "-" | "." | [0-9] | #xB7 | [#x0300-#x036F] | [#x203F-#x2040]
```

What do you think?